src/v/cluster/archival/ntp_archiver_service.cc

```diff
@@ -1606,8 +1606,21 @@ ntp_archiver::make_segment_index(
   retry_chain_logger& ctxlog,
   std::string_view index_path,
   ss::input_stream<char> stream) {
-    auto base_kafka_offset = model::offset_cast(
-      _parent.log()->from_log_offset(base_rp_offset));
+    std::exception_ptr eptr;
+    auto base_kafka_offset = [this, base_rp_offset, &eptr]() -> kafka::offset {
+        try {
+            return model::offset_cast(
+              _parent.log()->from_log_offset(base_rp_offset));
+        } catch (const std::runtime_error&) {
+            eptr = std::current_exception();
+            return kafka::offset{};
+        }
+    }();
+
+    if (eptr) {
+        co_await stream.close();
+        std::rethrow_exception(eptr);
```

Comment:
when this throws, the caller in `ntp_archiver::upload_segment` will destroy the other fanout stream `upload_stream`, and the same assertion will trigger?

Reply:
i don't see it. what am i missing?

```c++
    auto [upload_stream, indexing_stream] = input_stream_fanout<2>(
      std::move(stream),
      config::shard_local_cfg().storage_read_readahead_count());
```

IIUC if either `upload_stream` or `indexing_stream` are destroyed without being closed the assertion fires.

```c++
    auto tx_upload_started = maybe_upload_aborted_tx(
      path, std::move(tx_ranges), rtc);

    auto make_index_started = make_segment_index(
      meta.base_offset,
      meta.base_timestamp,
      ctxlog,
      index_path,
      std::move(indexing_stream));
```

And if `make_segment_index` throws, then your fix in this PR correctly handles `indexing_stream` (inside `make_segment_index`), but in this context, `upload_stream` will be destroyed?

Reply:
https://github.com/redpanda-data/redpanda/blob/29b610d90fade0cd430dfa5ec673810ee13a3f24/src/v/cluster/archival/ntp_archiver_service.cc#L1602-L1603

https://github.com/redpanda-data/redpanda/blob/29b610d90fade0cd430dfa5ec673810ee13a3f24/src/v/cluster/archival/ntp_archiver_service.cc#L1697-L1702

https://github.com/redpanda-data/redpanda/blob/29b610d90fade0cd430dfa5ec673810ee13a3f24/src/v/cluster/archival/ntp_archiver_service.cc#L1739-L1740

Reply:
> auto make_index_started = make_segment_index( 

oh ok i see now `make_segment_index` will return an exceptional future, of course!

Reply:
yup yup 👍